modules/rest-compatibility/src/main/java/org/elasticsearch/rest/action/document/RestTermVectorsActionV7.java

```diff
@@ -0,0 +1,84 @@
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.rest.action.document;
+
+import org.apache.logging.log4j.LogManager;
+import org.elasticsearch.Version;
+import org.elasticsearch.action.termvectors.TermVectorsRequest;
+import org.elasticsearch.client.node.NodeClient;
+import org.elasticsearch.common.logging.DeprecationLogger;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.compat.TypeConsumer;
+import org.elasticsearch.rest.RestRequest;
+import org.elasticsearch.rest.action.RestToXContentListener;
+
+import java.io.IOException;
+import java.util.List;
+
+import static org.elasticsearch.rest.RestRequest.Method.GET;
+import static org.elasticsearch.rest.RestRequest.Method.POST;
+
+public class RestTermVectorsActionV7 extends RestTermVectorsAction {
+    private static final DeprecationLogger deprecationLogger = new DeprecationLogger(LogManager.getLogger(RestTermVectorsAction.class));
+    public static final String TYPES_DEPRECATION_MESSAGE = "[types removal] " + "Specifying types in term vector requests is deprecated.";
+
+    @Override
+    public List<Route> routes() {
+        return List.of(
+            new Route(GET, "/{index}/_termvectors"),
+            new Route(POST, "/{index}/_termvectors"),
+            new Route(GET, "/{index}/_termvectors/{id}"),
+            new Route(POST, "/{index}/_termvectors/{id}"),
+            // Deprecated typed endpoints.
+            new Route(GET, "/{index}/{type}/_termvectors"),
+            new Route(POST, "/{index}/{type}/_termvectors"),
+            new Route(GET, "/{index}/{type}/{id}/_termvectors"),
+            new Route(POST, "/{index}/{type}/{id}/_termvectors")
+        );
+    }
+
+    @Override
+    public String compatibleWithVersion() {
+        return String.valueOf(Version.V_7_0_0.major);
+    }
+
+    @Override
+    public RestChannelConsumer prepareRequest(final RestRequest request, final NodeClient client) throws IOException {
+        TypeConsumer typeConsumer = new TypeConsumer(request,"_type");
+
+        TermVectorsRequest termVectorsRequest = new TermVectorsRequest(request.param("index"), request.param("id"));
+
+        if (request.hasContentOrSourceParam()) {
+            try (XContentParser parser = request.contentOrSourceParamParser()) {
+                TermVectorsRequest.parseRequest(termVectorsRequest, parser, typeConsumer);
+            }
+        }
+        readURIParameters(termVectorsRequest, request);
+
+        if (typeConsumer.hasTypes()) {
```

Comment:
If you move hasTypes a static method in a utility class, maybe we can push this logic (consume the type and log) there ?